File: src/main/java/club/someoneice/pineapple_delight/ItemInit.java
package club.someoneice.pineapple_delight;

import com.google.common.collect.Lists;
import net.minecraft.component.type.FoodComponent;
import net.minecraft.entity.LivingEntity;
import net.minecraft.entity.effect.StatusEffect;
import net.minecraft.entity.effect.StatusEffectInstance;
import net.minecraft.entity.effect.StatusEffects;
import net.minecraft.entity.player.PlayerEntity;
import net.minecraft.item.Item;
import net.minecraft.item.ItemStack;
import net.minecraft.item.Items;
import net.minecraft.registry.Registries;
import net.minecraft.registry.Registry;
import net.minecraft.registry.entry.RegistryEntry;
import net.minecraft.util.Identifier;
import net.minecraft.util.UseAction;
import net.minecraft.world.World;

import java.util.List;

public class ItemInit {
    public static final List<Item> ITEMS = Lists.newArrayList();

    public static final Item PINEAPPLE_ITEM = itemFoodBase("pineapple", 3, 0.5f, true, false, false);
    public static final Item PINEAPPLE_SIDE = itemFoodBase("pineapple_side", 1, 0.5f, true, false, true);
    public static final Item PINEAPPLE_PIE_SIDE = itemFoodBase("pineapple_pie_side", 3, 0.1f, false, false, false);
    public static final Item PINEAPPLE_JUICE = itemFoodDrink("pineapple_juice", StatusEffects.JUMP_BOOST);
    public static final Item PINEAPPLE_FRIED_RICE = itemFoodBase("pineapple_fried_rice", 12, 0.4f, false, false, false);
    public static final Item PINEAPPLE_MILK_SHAKE = itemFoodDrink("pineapple_milk_shake", StatusEffects.HEALTH_BOOST);
    public static final Item PINEAPPLE_ICE_CREAM = itemFoodDrink("pineapple_ice_cream", StatusEffects.SPEED);


    private static Item itemFoodBase(String name, int hunger, float saturation, boolean wolf, boolean fast, boolean alwaysEat) {
        var builder = new FoodComponent.Builder();
        builder.nutrition(hunger).saturationModifier(saturation);
        if (fast) builder.snack();
        if (alwaysEat) builder.alwaysEdible();
        return registry(new Item(new Item.Settings().food(builder.build())), name);
    }

    private static Item registry(Item item, String name) {
        Registry.register(Registries.ITEM, Identifier.of("pineapple_delight", name), item);
        ITEMS.add(item);
        return item;
    }
    
    private static Item itemFoodDrink(String name, RegistryEntry<StatusEffect> ... effects) {
        var builder = new FoodComponent.Builder();
        builder.nutrition(5).saturationModifier((float) 0.5);
        return registry(new ItemDrink(new Item.Settings().food(builder.build())).setReturnItem(Items.GLASS_BOTTLE).setEffect(effects), name);
    }

    private static class ItemDrink extends Item {
        ItemStack returnItem;
        RegistryEntry<StatusEffect>[] effects = null;

        public ItemDrink(Settings settings) {
            super(settings);
        }

        public ItemDrink setReturnItem(Item item) {
            this.returnItem = item.getDefaultStack();
            return this;
        }

        public ItemDrink setEffect(RegistryEntry<StatusEffect> ... effect) {
            this.effects = effect;
            return this;
        }

                                   @Override
        public UseAction getUseAction(ItemStack stack) {
            return UseAction.DRINK;
        }

        @Override
        public ItemStack finishUsing(ItemStack stack, World world, LivingEntity user) {
            super.finishUsing(stack, world, user);
            if (!returnItem.isEmpty()) ((PlayerEntity) user).giveItemStack(returnItem);
            if (effects != null) for (RegistryEntry<StatusEffect> effect : effects)
                user.addStatusEffect(new StatusEffectInstance(effect, 20 * 30, 1));

            return stack;
        }
    }
}

File: src/main/kotlin/club/someoneice/pineapple/ItemInit.kt
package club.someoneice.pineapple

import net.minecraft.block.Block
import net.minecraft.item.BlockItem
import net.minecraft.item.FoodComponent
import net.minecraft.item.Item
import net.minecraft.item.Item.Settings
import net.minecraft.util.Identifier
import net.minecraft.util.registry.Registry


object ItemInit {
    /* Foods Items */
    val PINEAPPLE: Item = ItemFoods(3, 0.5f, true, false, false)
    val PINEAPPLE_SIDE: Item = ItemFoods(1, 0.5f, true, false, true)
    val PINEAPPLE_PIE_SIDE: Item = ItemFoods(3, 0.5f, false, false, false)
    val PINEAPPLE_JUICE: Item = ItemFoods(5, 0.5f, false, false)
    val PINEAPPLE_FRIED_RICE: Item = ItemFoods(12, 0.4f, false, false, false)
    val PINEAPPLE_MILK_SHAKE: Item = ItemFoods(5, 0.5f, false, false)
    val PINEAPPLE_ICE_CREAM: Item = ItemFoods(5, 0.5f, false, false)

    /* Block Items */
    val PINEAPPLE_PIE = ItemBlockItems(BlockInit.PINEAPPLE_PIE)
    val PINEAPPLE_CROP = ItemBlockItems(BlockInit.PINEAPPLE_CROP)
    val PINEAPPLE_WILE_CROP = ItemBlockItems(BlockInit.WILD_PINEAPPLE_CROP)

    fun init() {
        Registry.register(Registry.ITEM, Identifier(modid, "pineapple"), PINEAPPLE)
        Registry.register(Registry.ITEM, Identifier(modid, "pineapple_side"), PINEAPPLE_SIDE)
        Registry.register(Registry.ITEM, Identifier(modid, "pineapple_pie_side"), PINEAPPLE_PIE_SIDE)
        Registry.register(Registry.ITEM, Identifier(modid, "pineapple_juice"), PINEAPPLE_JUICE)
        Registry.register(Registry.ITEM, Identifier(modid, "pineapple_fried_rice"), PINEAPPLE_FRIED_RICE)
        Registry.register(Registry.ITEM, Identifier(modid, "pineapple_milk_shake"), PINEAPPLE_MILK_SHAKE)
        Registry.register(Registry.ITEM, Identifier(modid, "pineapple_ice_cream"), PINEAPPLE_ICE_CREAM)

        Registry.register(Registry.ITEM, Identifier(modid, "pineapple_pie"), PINEAPPLE_PIE)
        Registry.register(Registry.ITEM, Identifier(modid, "pineapple_wild_crop"), PINEAPPLE_CROP)
        Registry.register(Registry.ITEM, Identifier(modid, "pineapple_crop"), PINEAPPLE_WILE_CROP)
    }


    fun ItemBlockItems(block: Block): Item {
        return BlockItem(block, Settings().group(PineappleGroup))
    }

    fun ItemFoods(hunger: Int, saturation: Float, wolf: Boolean, alwaysEat: Boolean, fast: Boolean, count: Int): Item {
        val set: Settings = Settings()
        val build: FoodComponent.Builder = FoodComponent.Builder()
        set.group(PineappleGroup)
        set.maxCount(count)

        build.hunger(hunger)
        build.saturationModifier(saturation)
        if (wolf) build.meat()
        if (fast) build.snack()
        if (alwaysEat) build.alwaysEdible()

        set.food(build.build())
        return Item(set)
    }

    fun ItemFoods(hunger: Int, saturation: Float, wolf: Boolean, alwaysEat: Boolean, fast: Boolean): Item {
        val set: Settings = Settings()
        val build: FoodComponent.Builder = FoodComponent.Builder()
        set.group(PineappleGroup)

        build.hunger(hunger)
        build.saturationModifier(saturation)
        if (wolf) build.meat()
        if (fast) build.snack()
        if (alwaysEat) build.alwaysEdible()

        set.food(build.build())
        return Item(set)
    }

    fun ItemFoods(hunger: Int, saturation: Float, alwaysEat: Boolean, fast: Boolean): Item {
        val set: Settings = Settings()
        val build: FoodComponent.Builder = FoodComponent.Builder()
        set.group(PineappleGroup)

        build.hunger(hunger)
        build.saturationModifier(saturation)
        if (fast) build.snack()
        if (alwaysEat) build.alwaysEdible()

        set.food(build.build())
        return DrinkItem(set)
    }
}